File: src/main/java/com/neutralplasma/virtusbot/storage/TicketStorage.java
package com.neutralplasma.virtusbot.storage;


import com.neutralplasma.virtusbot.storage.dataStorage.SQL;
import com.neutralplasma.virtusbot.storage.dataStorage.StorageHandler;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class TicketStorage {
    private StorageHandler sql;
    private final String tableName = "TicketData";

    public TicketStorage(StorageHandler sql){
        this.sql = sql;
    }

    public void setup(){
        try {
            sql.createTable(tableName, "userID TEXT, channelID TEXT");
        }catch (Exception error){
            error.printStackTrace();
        }
    }
    public void writeSettings(String userID, String channelID){
        TicketInfo info = new TicketInfo(userID, channelID);
        try {
            addTicket(tableName, info);
        }catch (Exception error){
            error.printStackTrace();
        }
    }

    public void deleteTicket(String userID, String channelID){
        TicketInfo info = new TicketInfo(userID, channelID);
        try {
            removeTicket(tableName, info);
        }catch (SQLException sqlerror){
            sqlerror.printStackTrace();
        }
    }

    public TicketInfo getTicket(String channelID){
        try {
            if (!channelID.isEmpty()) {
                TicketInfo ticketid = getTicketbyChannel(tableName, channelID);
                if (ticketid != null) {
                    return ticketid;
                }
            }

            return null;
        }catch (Exception error){
            error.printStackTrace();
        }
        return null;
    }

    public String getTicketID(String userID){
        try {
            if (!userID.isEmpty()) {
                TicketInfo ticketid = getTicket(tableName, userID);
                if (ticketid != null) {
                    return ticketid.getChannelID();
                }
            }
            return null;
        }catch (Exception error){
            error.printStackTrace();
        }
        return null;
    }

    public TicketInfo getTicket(String tablename, String usedid) throws SQLException{
        try(Connection connection = sql.getConnection()) {
            String statement = "SELECT * FROM " + tablename + " WHERE userID = ?";
            try(PreparedStatement preparedStatement = connection.prepareStatement(statement)){
                preparedStatement.setString(1, usedid);
                ResultSet resultSet = preparedStatement.executeQuery();
                while(resultSet.next()){
                    TicketInfo ticket = new TicketInfo(resultSet.getString("userID"), resultSet.getString("channelID"));
                    return ticket;
                }
            }
        }
        return null;
    }

    public TicketInfo getTicketbyChannel(String tablename, String channelID) throws SQLException{
        try(Connection connection = sql.getConnection()) {
            String statement = "SELECT * FROM " + tablename + " WHERE channelID = ?";
            try(PreparedStatement preparedStatement = connection.prepareStatement(statement)){
                preparedStatement.setString(1, channelID);
                ResultSet resultSet = preparedStatement.executeQuery();
                while(resultSet.next()){
                    TicketInfo ticket = new TicketInfo(resultSet.getString("userID"), resultSet.getString("channelID"));
                    return ticket;
                }
            }
        }
        return null;
    }

    public boolean addTicket(String tablename, TicketInfo info) throws SQLException{
        if(getTicket(tablename, info.userid) != null){
            return false;
        }
        try(Connection connection = sql.getConnection()){
            String statement = "INSERT INTO " +
                    " " + tablename + " (userID, ChannelID) " +
                    "VALUES (?, ?)";
            try(PreparedStatement preparedStatement = connection.prepareStatement(statement)){
                preparedStatement.setString(1, info.userid);
                preparedStatement.setString(2, info.channelID);
                preparedStatement.execute();
                return true;
            }
        }
    }

    public boolean removeTicket(String tablename, TicketInfo info) throws SQLException {
        if(getTicket(tablename, info.userid) != null) {
            try (Connection connection = sql.getConnection()) {
                String statement = "DELETE FROM " + tablename + " WHERE userID = ? AND channelID = ?";
                try (PreparedStatement preparedStatement = connection.prepareStatement(statement)) {
                    preparedStatement.setString(1, info.userid);
                    preparedStatement.setString(2, info.channelID);
                    preparedStatement.execute();
                }

            }
        }

        return false;
    }

}

File: src/main/java/com/neutralplasma/virtusbot/storage/ticket/TicketStorage.kt
package com.neutralplasma.virtusbot.storage.ticket

import com.neutralplasma.virtusbot.storage.dataStorage.StorageHandler
import java.sql.SQLException
import kotlin.jvm.Throws

class TicketStorage(private val sql: StorageHandler) {
    private val tableName = "TicketStorage"
    fun setup() {
        try {
            sql.createTable(tableName, "userID TEXT, channelID TEXT, guildID TEXT, closed TINYINT")
        } catch (error: Exception) {
            error.printStackTrace()
        }
    }


    fun createTicket(info: TicketInfo): Boolean{
        if(getTicket(info) == null){
            sql.connection.use { connection ->
                val statement = "INSERT INTO $tableName (userID, ChannelID, guildID, closed) " +
                        "VALUES (?, ?, ?, ?)"
                connection?.prepareStatement(statement)?.use {
                    it.setString(1, info.userID)
                    it.setString(2, info.channelID)
                    it.setString(3, info.guildID)
                    it.setInt(4, info.closed)
                    it.execute()
                    return true
                }
            }
        }
        return false
    }

    fun deleteTicket(info: TicketInfo): Boolean{
        if(getTicket(info) != null){
            sql.connection.use {
                val statement = "DELETE FROM $tableName WHERE userID = ? AND channelID = ? AND guildID = ? AND closed = ?"
                it?.prepareStatement(statement)?.use { st ->
                    st.setString(1, info.userID)
                    st.setString(2, info.channelID)
                    st.setString(3, info.guildID)
                    st.setInt(4, info.closed)
                    st.execute()
                }
            }
        }
        return false
    }

    fun getTicket(info: TicketInfo): TicketInfo?{
        sql.connection.use {
            val statement = "SELECT * FROM $tableName WHERE userID = ? AND channelID = ? AND guildID = ? AND closed = ?"
            it?.prepareStatement(statement)?.use {st ->
                st.setString(1, info.userID)
                st.setString(2, info.channelID)
                st.setString(3, info.guildID)
                st.setInt(4, info.closed)
                val resultSet = st.executeQuery()
                while (resultSet.next()) {
                    return TicketInfo(
                        resultSet.getString("userID"),
                        resultSet.getString("channelID"),
                        resultSet.getInt("closed"),
                        resultSet.getString("guildID"))
                }
            }
        }
        return null
    }
    fun getTicketGuild(guildID: String, userID: String): TicketInfo?{
        sql.connection.use {
            val statement = "SELECT * FROM $tableName WHERE userID = ? AND guildID = ?"
            it?.prepareStatement(statement)?.use {st ->
                st.setString(1, userID)
                st.setString(2, guildID)
                val resultSet = st.executeQuery()
                while (resultSet.next()) {
                    return TicketInfo(
                        resultSet.getString("userID"),
                        resultSet.getString("channelID"),
                        resultSet.getInt("closed"),
                        resultSet.getString("guildID"))
                }
            }
        }
        return null
    }

    fun getTicketGuild(guildID: String, userID: String, closed: Int): TicketInfo?{
        sql.connection.use {
            val statement = "SELECT * FROM $tableName WHERE userID = ? AND guildID = ? AND closed = ?"
            it?.prepareStatement(statement)?.use {st ->
                st.setString(1, userID)
                st.setString(2, guildID)
                st.setInt(3, closed)
                val resultSet = st.executeQuery()
                while (resultSet.next()) {
                    return TicketInfo(
                        resultSet.getString("userID"),
                        resultSet.getString("channelID"),
                        resultSet.getInt("closed"),
                        resultSet.getString("guildID"))
                }
            }
        }
        return null
    }

    fun getTicketChannel(channelID: String, userID: String): TicketInfo?{
        sql.connection.use {
            val statement = "SELECT * FROM $tableName WHERE userID = ? AND channelID = ?"
            it?.prepareStatement(statement)?.use {st ->
                st.setString(1, userID)
                st.setString(2, channelID)
                val resultSet = st.executeQuery()
                while (resultSet.next()) {
                    return TicketInfo(
                        resultSet.getString("userID"),
                        resultSet.getString("channelID"),
                        resultSet.getInt("closed"),
                        resultSet.getString("guildID"))
                }
            }
        }
        return null
    }

    fun getTicketChannel(channelID: String): TicketInfo?{
        sql.connection.use {
            val statement = "SELECT * FROM $tableName WHERE channelID = ?"
            it?.prepareStatement(statement)?.use {st ->
                st.setString(1, channelID)
                val resultSet = st.executeQuery()
                while (resultSet.next()) {
                    return TicketInfo(
                        resultSet.getString("userID"),
                        resultSet.getString("channelID"),
                        resultSet.getInt("closed"),
                        resultSet.getString("guildID"))
                }
            }
        }
        return null
    }

    fun closeTicket(info: TicketInfo){
        sql.connection.use {
            val statement = "UPDATE $tableName SET closed = ? WHERE channelID = ? AND userID = ? AND guildID = ?"
            it?.prepareStatement(statement)?.use { st ->
                st.setInt(1, 1)
                st.setString(2, info.channelID)
                st.setString(3, info.userID)
                st.setString(4, info.guildID)
                st.execute()
            }
        }
    }
}